File: ProblemSolving/Python/arrays/hash_map_set.py
"""
This script contains my solutions for gfg problems that can be solved mainly using a hashmap or a set
"""

from typing import List
from collections import defaultdict, deque, Counter

# https://www.geeksforgeeks.org/problems/smallest-positive-missing-number-1587115621/1?page=1&category=Arrays&difficulty=Medium&status=unsolved&sortBy=submissions
def missingNumber(arr: List[int]) -> int:
    # basically convert all positive numbers into a set and iterate from 1 
    pos_nums_set = set()
    max_element = -float("inf")
    
    for v in arr:
        if v >= 1:
            pos_nums_set.add(v)
        max_element = max(max_element, v)

    # if all element are non-positive, then return 1
    if max_element < 1:
        return 1

    # otherwise iterate from "1" to "max_element + 1", the last number if a guaranteed miss
    for i in range(1, max_element + 2):
        if i not in pos_nums_set:
            return i


# https://www.geeksforgeeks.org/problems/array-pair-sum-divisibility-problem3257/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card

# 1st check: must have an even number of element 
# save the count of elements with each reminder: {i: {# v | v % k = i }}

def canPair(nums: List[int], k: int):
    if len(nums) % 2 == 1:
        return False
    
    k_nums = [v % k for v in nums]
    
    k_map = dict([(v, 0) for v in k_nums])

    for v in k_nums:
        k_map[v] += 1

    # element divisible by "k" are associated together: need an even count of such numbers
    if 0 in k_map and  k_map[0] % 2 == 1:
        return False

    # if k is even then numbers with reminder k // 2 are grouped together, need an even count of such numbers
    if (k % 2 == 0) and (k // 2) in k_map and (k_map[k // 2] % 2 == 1):
        return False
    
    # for other reminders, then r and k - r need to be grouped together, the total count for each of those reminders must be the same
    for rem, _ in k_map.items():
        other_rem = (k - rem) % k
        
        if (other_rem not in k_map) or (k_map[other_rem] != k_map[rem]):
            return False
        
    return True


# going through few problems from this list: 
# https://www.geeksforgeeks.org/hashing-data-structure/?ref=lbp
# some of them are just dumb...

# somehow this problem has an accuracy of 30%
# https://www.geeksforgeeks.org/problems/incomplete-array3859/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card
def countElements(n: int, a: List[int]) -> int:
    min_val, max_val = a[0], a[0]
    for v in a: 
        min_val = min(min_val, v)
        max_val = max(max_val, v)

    return max_val - min_val + 1 - len(set(a))


"""
https://www.geeksforgeeks.org/problems/count-distinct-elements-in-every-window/1
"""
def countDistinct(arr, n, k):
    counter = {}
    for v in arr[:k]:
        if v not in counter:
            counter[v] = 0
        counter[v] += 1
    
    res = [0 for _ in range(n - k + 1)]
    res[0] = len(counter)

    for i in range(k, n):
        val_remove = arr[i - k]
        val_add = arr[i]

        # add val_add
        if val_add not in counter:
            counter[val_add] = 0
        
        counter[val_add] += 1
        
        counter[val_remove] -= 1
        if counter[val_remove] == 0:
            del(counter[val_remove])

        res[i - k + 1] = len(counter)

    return res


# https://www.geeksforgeeks.org/problems/find-rectangle-with-corners-as-1--141631/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card 
def ValidCorner(matrix: List[List[int]]): 
    # iterate thorugh
    pass


# a problem from neetcode.io: https://neetcode.io/problems/string-encode-and-decode
from typing import List
def encode(strs: List[str]) -> str:
    if len(strs) == 0:
        return ""
    
    encoded_list = ["_".join([str(ord(c)) for c in s]) if len(s) > 0 else "" for s in strs]
    return "*" + "*".join(encoded_list)
    
def decode(string: str) -> List[str]:
    if len(string) == 0:
        return []
    
    string = string[1:]
    l = string.split("*")
    return ["".join([chr(int(c)) for c in s.split("_")]) if len(s) > 0 else "" for s in l ]


"""
https://www.geeksforgeeks.org/problems/largest-subarray-of-0s-and-1s/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=bottom_sticky_on_article
"""
def maxLen(arr, N=None):
    if N is None:
        N = len(arr)
    # my solution combines the idea of hashing and PrefixSum array
    count0 = 0
    count1 = 0
    prefixSum = [0 for _ in arr]
    max_len = 0
    for index, val in enumerate(arr):
        count0 += int(val == 0)
        count1 += int(val == 1) 
        prefixSum[index] = count1 - count0
        
        if count0 == count1:
            max_len = index + 1

    if max_len == N:
        return N

    pfs_indices = defaultdict(lambda:[])

    for index, val in enumerate(prefixSum):
        pfs_indices[val].append(index)
    
    for _, v in pfs_indices.items():
        max_len = max(max_len, abs(v[0] - v[-1]))
    
    return max_len


"""
https://www.geeksforgeeks.org/problems/count-pairs-with-given-sum5022/1
"""
def getPairsCount(arr, target):
    # get the frequencies of each element in the arrya
    freqs = defaultdict(lambda : 0)
    count = 0

    for val in arr:
        freqs[val] += 1

    for val in arr:
        new_t = target - val
        if new_t == val:
            count += (freqs[val] - 1) / 2
        else:
            count += freqs[new_t] / 2
    
    return int(count)

"""
https://www.geeksforgeeks.org/problems/first-repeating-element4018/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=bottom_sticky_on_article
"""
def firstRepeated(arr):
    freqs = {}
    q = deque()
    for index, val in enumerate(arr):
        if val not in freqs:
            freqs[val] = 0
            q.append((val, index))
        freqs[val] += 1
    
    while len(q) > 0 and freqs[q[0][0]] == 1:
        q.popleft()
    
    if len(q) == 0:
        return -1

    return q[0][1] + 1


# Although this problem is already solved, I see it as a great example of combining the Dict and Queue data structures: 
"""
https://www.geeksforgeeks.org/given-a-string-find-its-first-non-repeating-character/#efficient-approach-1-using-hash-map-on-time-and-o1-auxiliary-space
"""

# let's start with the first efficient solution
def first_not_repeat_char_2_traversals(string: str) -> int:
    # the first natural idea it to compute the frequency of each character: first iteration
    # iterate again: consult the frequency of each char, if it is 1: return else, return -1 at the end...
    freqs = {}
    for c in string:
        if c not in freqs:
            freqs[c] = 0
        freqs[c] += 1
    
    for c in string:
        if freqs[c] == 1:
            return c
    return -1

# this is a better solution with only on string traversal
def first_not_repeat_char_1_traversal(string: str) -> int:
    # my main line of thought before solving this question was: assuming at each step I am saving the current first character 
    # occuring once so far in the string, how can I get the next char with only one occurrence once the current save char is repeated
    
    ## well you've guessed it a queue !!!

    freqs = {}
    # the queue point here is that queue should save only characters with frequency 1 (in order)
    queue = deque()

    for c in string: 
        if c not in freqs:
            queue.append(c)
            freqs[c] = 0
        
        freqs[c] += 1
        # the two lines below ensure that the queue contains
        # only characters that appear exactly once
        # + the FIFO nature of the queue ensures that the first element in the queue appears first in the STRING !!! 
        while len(queue) > 0 and freqs[queue[0]] != 1:
            queue.pop()
    
    return queue[0] if len(queue) else -1


"""
https://www.geeksforgeeks.org/problems/longest-consecutive-subsequence2449/1
"""
def findLongestConseqSubseq(array, N = None):
    if N is None:
        N = len(array)
    arr = set(array)
    visited = dict([(a, False) for a in arr])

    max_len = 0
    for v in arr:
        if visited[v]:
            continue 
        
        # v has not been visited below
        start = v - 1
        while start in arr:
            start = start - 1
        
        start = start + 1
        counter = 0
        while start in arr:
            visited[start] = True
            start += 1
            counter += 1

        max_len = max(counter, max_len)

    return max_len


# https://www.geeksforgeeks.org/problems/kth-distance3757/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card
def checkDuplicatesWithinK(arr: List[int], k: int):
    # basically we need to find if there is an element for which there exists i, j such that arr[i] == arr[j] and abs(i - j) <= k

    # using a hashmap is probably the first idea that comes to mind...    
    # associate each number with the last occurrence if a new occurrence is detected, basically check the distance between the indices

    val_indices = {}

    for i, v in enumerate(arr):
        if v in val_indices:
            if i - val_indices[v] <= k:
                return True
        # update val_indices anyway
        val_indices[v] = i
    
    return False


# this is a very simple problem: 
# https://www.geeksforgeeks.org/problems/remove-duplicates-in-small-prime-array/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card 
def removeDuplicates(arr: List[int]):
    arr_set = set(arr)
    n = len(arr_set)
    
    arr_set.clear()
    
    new_arr = [0 for _ in range(n)]
    counter = 0

    for v in arr:
        if v not in arr_set:
            arr_set.add(v)
            new_arr[counter] = v
            counter += 1
    
    return new_arr
    

# the simplest DSA question ever...
def isSubset( a1, a2, n=None, m=None):
    c1, c2 = Counter(a1), Counter(a2)
    for k, v in c2.items():
        if k not in c1:
            return False
        if v > c1[k]:
            return False
    return True


# this question for some reason has an 18% accuracy percetange (I mean, it is a waste of time to do, if one reaches a certain level)
def findDuplicates(arr: List[int]):
    d = {}
    count = 0
    for v in arr:
        if v not in d:
            d[v] = 0

        if d[v] == 1:
            count += 1
        
        d[v] += 1

    res = [0 for _ in range(count)]

    i = 0
    for k, freq in d.items():
        if freq > 1: 
            res[i] = k
            i += 1
    
    return res



def findTwoElement(arr: List[int]):
    n = len(arr)
    arr_set = set(arr)
    missing = 0 

    for v in range(1, n + 1):
        if v not in arr_set:
            missing = v
            break
    
    # find the repeating number
    diff = (n * (n + 1)) // 2 - sum(arr)
    repeating = missing - diff
    return repeating, missing

File: ProblemSolving/Cpp/arrays/hash_map_set.cpp
/*
This file contains my solutions for gfg problems that can be solved mainly using a hashmap or a set
*/ 

# include "hash_map_set.h"

bool map_subset_map(std::unordered_map<char, vi> m1, std::unordered_map<char, int> m2) {

    // check if m2 is a subset of m1
    auto it = m2.begin();
    while (it != m2.end()) {
        char key = it -> first;
        // check if the key is in m1

        auto m1_key = m1.find(key);
        if (m1_key == m1.end()) {
            return false;
        }

        // check the occurrence
        if ((m1_key -> second).size() < it -> second){
            return false;
        }
        it ++;
    }
    return true;
}


// the main goal of solving this problem is to refresh the C++ syntax in my head

std::vector<int> removeDuplicate(std::vector<int>& arr) {
    // the idea is very 
    std::set arr_set = std::set<int>();

    vi new_arr = vi();

    for (const int v : arr) {
        if (arr_set.find(v) == arr_set.end()) {
        // if (! arr_set.contains(v) ) {
            new_arr.push_back(v);
            arr_set.insert(v);
        }
    }
    return new_arr;
}


// std::string smallestWindow (std::string& s, std::string& p) {

//     // consider the case where p is of length 1 independently
//     if (p.size() == 1) {
//         if (s.find(p) != std::string::npos) {
//             return p;
//         }
//         return "-1";
//     }


//     // let's build p as a dictionary
//     std::unordered_map<char, int> p_map{};
//     for (char c: p) {
//         if (p_map.find(c) == p_map.end()) {
//             p_map[c] = 1;
//         } else {
//             p_map[c] += 1;
//         }
//     }

//     int best_start = -1, best_end = static_cast<int>(s.size());

//     // let's first find a good start
//     int start = 0;
//     while (p_map.find(s[start]) == p_map.end()) {
//         start++;
//     }
//     // at this point of the program we know that p has at least two elements
//     // hence end should be at least start + 1

//     int end = start + 1;

//     // at this point we know that s[start] is common with 'p'
//     std::unordered_map<char, vi> segment_map = {{s[start], {start}}};

//     int n = static_cast<int> (s.size());

//     while (end < n) {
//         // check if s[end] is in the p_map
//         if (p_map.find(s[end]) != p_map.end()) {
//             // append the value of 'end' to the vector associated with s[end]
//             if (segment_map.find(s[end]) == segment_map.end()) {
//                 segment_map[s[end]] = {end};
//             } else {
//                 segment_map[s[end]].push_back(end);
//             }
//             // check if the segment contains 'p'
//             bool segment_contains_p = map_subset_map(segment_map, p_map);

//             if (segment_contains_p) {
//                 if (end - start < best_end - best_start) {
//                     // update the best segment
//                     best_start = start;
//                     best_end = end;
//                 }
//                 // time to update the 'start' variable:
//                 // first remove it from the segment
//                 // choose start as the smallest value in the current segment
//                 auto it = segment_map.begin();

//                 int index_to_remove = start;
//                 char char_to_remove = s[start];

//                 while (it != segment_map.end()) {
//                     int size = static_cast<int>((it->second).size());
//                     int needed_size = p_map[it->first];

//                     if (size == needed_size) {
//                         it ++ ;
//                         continue;
//                     }

//                     int new_index = (it->second)[size - needed_size - 1];

//                     if (new_index > index_to_remove) {
//                         index_to_remove = std::max(index_to_remove, new_index);
//                         char_to_remove = it->first;
//                     }
//                     it ++ ;
//                 }

//                 int needed_size = p_map[char_to_remove], size = segment_map[char_to_remove].size();
//                 auto b = segment_map[char_to_remove].begin(), e = b + size - needed_size;
//                 segment_map[char_to_remove].erase(b, e);

//                 int new_start = end;
//                 it = segment_map.begin();
//                 // find the new start
//                 while (it != segment_map.end()) {
//                     new_start = std::min((it->second)[0], new_start);
//                     it ++;
//                 }

//                 start = new_start;
//                 std::string r = "";
//             }

//         }

//         end++;
//     }

//     bool segment_contains_p = map_subset_map(segment_map, p_map);
//     if (segment_contains_p) {
//         if (end - start < best_end - best_start) {
//             // update the best segment
//             best_start = start;
//             best_end = end;
//         }
//     }
//     // make sure to consider the case where a match is not found
//     if (best_start == -1) {
//         return "-1";
//     }

//     return s.substr(best_start, best_end - best_start + 1);
// }